### PR TITLE
 Precompile events - Fee Config

### PR DESCRIPTION
**Pull request by abi87**
## Why this should be merged

Adds event to set fee config function. 

## How this works

Adds new Event to ABI and regenerated with the precompile generation tool.

`event FeeConfigChanged(address indexed sender, FeeConfig oldFeeConfig, FeeConfig newFeeConfig);`

Increases total gas cost by 45221 (base event gas, reading old fee config from statedb, 2 topics gas, 512-byte unindexed data)

## How this was tested

Added UT and hardhat tests.

## How is this documented

Will be documented with other Durango changes

### CODE REVIEW

File: precompile/contracts/feemanager/contract.go [hunk 1]
```diff
@@ -242,6 +249,34 @@ func setFeeConfig(accessibleState contract.AccessibleState, caller common.Addres
...
+		topics, data, err := PackFeeConfigChangedEvent(
+			caller,
+			oldConfig,
+			feeConfig,
+		)
+		if err != nil {
+			return nil, remainingGas, err
+		}
+
+		logDataGasCost := uint64(len(data)) * contract.LogDataGas
```
Comment (aaronbuchwald) [inline precompile/contracts/feemanager/contract.go:267]: Could we subtract `GetFeeConfigGasCost` at the top here and move the entire calculation of the gas cost due to the produced log into one place, so that it's not split up?

File: precompile/contracts/feemanager/contract_test.go [hunk 2]
```diff
@@ -258,3 +332,14 @@ func TestFeeManager(t *testing.T) {
...
 }
+
+func getGasCost(oldFeeConfig commontype.FeeConfig, feeConfig commontype.FeeConfig) uint64 {
+	// address doesn't matter for gas cost
+	_, data, err := PackFeeConfigChangedEvent(common.Address{}, oldFeeConfig, feeConfig)
+	if err != nil {
+		panic(err)
+	}
+
+	logGasCost := GetFeeConfigGasCost + FeeConfigChangedEventBaseGasCost + contract.LogDataGas*uint64(len(data))
```
Comment (aaronbuchwald) [inline precompile/contracts/feemanager/contract_test.go:343]: All of the fields of the event are a fixed size, so this should be a constant value regardless of the inputs

File: precompile/contracts/feemanager/contract_test.go [hunk 3]
```diff
@@ -248,6 +250,78 @@ var (
 			ReadOnly:    false,
 			ExpectedErr: vmerrs.ErrOutOfGas.Error(),
 		},
+		"set config doesn't log before DUpgrade": {
```
Comment (aaronbuchwald) [inline precompile/contracts/feemanager/contract_test.go:253]: same comment from other event PRs, doesn't creating a before and after + modifying the previously existing test create a duplicate of the same case?

File: precompile/contracts/feemanager/event.go [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+type ChangeFeeConfigEventData struct {
+	GasLimit                 *big.Int
+	TargetBlockRate          *big.Int
+	MinBaseFee               *big.Int
+	TargetGas                *big.Int
+	BaseFeeChangeDenominator *big.Int
+	MinBlockGasCost          *big.Int
+	MaxBlockGasCost          *big.Int
+	BlockGasCostStep         *big.Int
+}
```
Comment (aaronbuchwald) [inline precompile/contracts/feemanager/event.go:27]: Could we add a comment to explain why we need the intermediate type between `commontype.FeeConfig` and `ChangeFeeConfigEventData` ?

**Review by aaronbuchwald — CHANGES_REQUESTED** [review]

File: precompile/contracts/feemanager/event.go (hunk 4 above)
Comment (ceyonur) [reply precompile/contracts/feemanager/event.go:27]: I think we can just unexport it and return commontype.FeeConfig instead.

File: precompile/contracts/feemanager/contract_test.go (hunk 2 above)
Comment (ceyonur) [reply precompile/contracts/feemanager/contract_test.go]: yep, changed them

File: precompile/contracts/feemanager/contract.go (hunk 1 above)
Comment (ceyonur) [reply precompile/contracts/feemanager/contract.go]: moved

File: precompile/contracts/feemanager/contract_test.go (hunk 3 above)
Comment (ceyonur) [reply precompile/contracts/feemanager/contract_test.go:278]: I moved after hook to existing test.

File: contracts/test/fee_manager.ts [hunk 5]
```diff
@@ -4,6 +4,10 @@
 import { expect } from "chai"
 import { ethers } from "hardhat"
 import { test } from "./utils"
+import { SignerWithAddress } from "@nomiclabs/hardhat-ethers/signers";
+import {
+  Contract,
+} from "ethers"
```
Comment (darioush) [inline contracts/test/fee_manager.ts:10]:
```suggestion
import { Contract } from "ethers"
```

File: contracts/test/fee_manager.ts [hunk 6]
```diff
@@ -53,3 +57,66 @@ describe("ExampleFeeManager", function () {
     "step_lowerMinFeeByOne",
   ])
 })
+const C_FEES = {
```
Comment (darioush) [inline contracts/test/fee_manager.ts:60]:
```suggestion

const C_FEES = {
```

File: precompile/contracts/feemanager/contract_test.go [hunk 7]
```diff
@@ -67,12 +77,27 @@ var (
...
+				topics := logsTopics[0]
+				require.Len(t, topics, 2)
+				require.Equal(t, FeeManagerABI.Events["FeeConfigChanged"].ID, topics[0])
+				require.Equal(t, allowlist.TestEnabledAddr.Hash(), topics[1])
+
+				logData := logsData[0]
+				oldFeeConfig, resFeeConfig, err := UnpackFeeConfigChangedEventData(logData)
+				require.NoError(t, err)
+				require.True(t, zeroFeeConfig.Equal(&oldFeeConfig), "expected %v, got %v", zeroFeeConfig, oldFeeConfig)
+				require.True(t, testFeeConfig.Equal(&resFeeConfig), "expected %v, got %v", testFeeConfig, resFeeConfig)
```
Comment (darioush) [inline precompile/contracts/feemanager/contract_test.go:100]: could we abstract this to a helper function instead of repeating it?

Comment (ceyonur) [reply precompile/contracts/feemanager/contract_test.go]: good idea 👍

**Review by darioush — APPROVED** [review]

File: precompile/allowlist/allowlist.go [hunk 8]
```diff
@@ -165,11 +165,26 @@ func CreateAllowListPrecompile(precompileAddr common.Address) contract.StatefulP
...
+		case ManagerRole.GetSetterFunctionName():
+			fn = contract.NewStatefulPrecompileFunctionWithActivator(method.ID, createAllowListRoleSetter(precompileAddr, ManagerRole), contract.IsDUpgradeActivated)
+		default:
+			panic(fmt.Sprintf("unexpected method name: %s", name))
+		}
+		functions = append(functions, fn)
+	}
 
-	return []*contract.StatefulPrecompileFunction{setAdmin, setManager, setEnabled, setNone, read}
+	return functions
```
Comment (aaronbuchwald) [inline precompile/allowlist/allowlist.go:189]: What's the reason for this change?

Comment (ceyonur) [reply precompile/allowlist/allowlist.go:189]: you might be looking at a merge commit.

**Review by aaronbuchwald — APPROVED** [review]
